fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/services/AbstractService.java

```diff
@@ -290,15 +290,17 @@ protected void lockArchivalGroupResource(final Transaction tx,
         }
     }
 
-    protected void lockArchivalGroupResourceFromParent(final Transaction tx,
-                                                       final PersistentStorageSession pSession,
-                                                       final FedoraId parentId) {
+    protected void lockParent(final Transaction tx,
+                              final PersistentStorageSession pSession,
+                              final FedoraId parentId) {
         if (parentId != null && !parentId.isRepositoryRoot()) {
             final var parentHeaders = pSession.getHeaders(parentId, null);
-            if (parentHeaders.isArchivalGroup()) {
-                tx.lockResource(parentId);
-            } else if (parentHeaders.getArchivalGroupId() != null) {
+            if (parentHeaders.getArchivalGroupId() != null) {
+                // parent is part of an AG, so lock the AG
                 tx.lockResource(parentHeaders.getArchivalGroupId());
+            } else {
+                // otherwise lock the parent.
+                tx.lockResource(parentId);
```

Comment:
Is this right? Won't this lock all atomic container parents?

Reply:
Right, so this is only used in the `CreateResourceServiceImpl` for creating a new resource and in the `WebacAclServiceImpl` also for creating a new ACL. Because of that I assumed we wouldn't want anyone... mostly deleting the parent while a new child is being created.

Reply:
Hmm... It's not immediately clear to me what would happen in the existing code if you concurrently added a child while deleting its parent. I would hope it would fail, but I'm not sure where. I am a little hesitant to lock the parent in these cases because it means that you can't concurrently add children to a container, which used to only be a limitation of AGs. However, we might not have a choice if the behavior change here is actually fixing a bug.

If we do need to lock, perhaps we need a followup ticket for different kinds of locks? That is to say, locking the parent so that it can't be deleted, but allowing multiple sessions to concurrently add children.

Reply:
Test added